src/System.CommandLine/CommandLineConfiguration.cs

```diff
@@ -20,7 +21,8 @@ public class CommandLineConfiguration
             bool enablePositionalOptions = false,
             ValidationMessages validationMessages = null,
             ResponseFileHandling responseFileHandling = ResponseFileHandling.ParseArgsAsLineSeparated,
-            IReadOnlyCollection<InvocationMiddleware> middlewarePipeline = null)
+            IReadOnlyCollection<InvocationMiddleware> middlewarePipeline = null,
+            Func<InvocationContext, IHelpBuilder> helpProvider = null)
```

Comment:
This should probably eventually be something a bit more specific, like an interface. @KathleenDollard is working on an app model including providers for this and other services.

Reply:
I think that would be a good idea. That is certainly pushing in the same direction that `Microsoft.Extensions.DependencyInjection.Abstractions` uses with factory interfaces that provide the actual services. 
I also considered that rather than making the creation of the `IHelpBuilder` lazy, it might be simpler to just take in an instance and pass both the `InvocationContext` to it rather than just the command.

Reply:
You can get to everything from the `InvocationContext` so it should add some flexibility relative to just passing the command. For example, things like this: https://github.com/dotnet/command-line-api/issues/358.